contracts/prize-pool/barnbridge/BarnPrizePool.sol

```diff
@@ -0,0 +1,126 @@
+        BarnInterface _barn,
+        BarnRewardsInterface _rewards,
+        IERC20Upgradeable _bond
+    ) public initializer {
+        PrizePool.initialize(
+            _reserveRegistry,
+            _controlledTokens,
+            _maxExitFeeMantissa,
+            _maxTimelockDuration
+        );
+        barn = _barn;
+        rewards = _rewards;
+        bond = _bond;
+
+        emit BarnPrizePoolInitialized(address(barn));
+    }
+
+    /// @dev Gets the balance of the underlying assets held by the Yield Service
+    /// @return The underlying balance of asset tokens
+    function _balance() internal override returns (uint256) {
+        uint256 balance = barn.balanceOf(address(this));
+        uint256 owed = rewards.owed(address(this));
+        return balance.add(owed);
+    }
+
+    /// @dev Allows a user to supply asset tokens in exchange for yield-bearing tokens
+    /// to be held in escrow by the Yield Service
+    function _supply(uint256 amount) internal override {
+        IERC20Upgradeable bondToken = _token();
+        bondToken.approve(address(barn), bondToken.totalSupply());
+        barn.deposit(amount);
+    }
+
+    /// @dev The external token cannot be yDai or Dai
+    /// @param _externalToken The address of the token to check
+    /// @return True if the token may be awarded, false otherwise
+    function _canAwardExternal(address _externalToken)
+        internal
+        view
+        override
+        returns (bool)
+    {
+        return
+            _externalToken != address(barn) &&
+            _externalToken != address(barn.token());
+    }
+
+    /// @dev Allows a user to redeem yield-bearing tokens in exchange for the underlying
+    /// asset tokens held in escrow by the Yield Service
+    /// @param amount The amount of underlying tokens to be redeemed
+    /// @return The actual amount of tokens transferred
+    function _redeem(uint256 amount) internal override returns (uint256) {
+        require(_balance() >= amount, "BarnPrizePool/insuff-liquidity");
+        IERC20Upgradeable token = _token();
+
+        uint256 diff = 0;
+        uint256 amountToClaim = rewards.owed(address(this));
+
+        /// Check if there is anything to claim from the rewards
+        if (amountToClaim > 0) {
+            rewards.claim();
+        }
+
+        uint256 currentBalance = token.balanceOf(address(this));
+
+        /// If current bond balance is enough, deposit the difference back to Barn
+        if (currentBalance >= amount) {
+            diff = currentBalance.sub(amount);
+            barn.deposit(diff);
+        }
+
+        /// If current bond balance is not enough, try to withdraw the difference from Barn
+        if (currentBalance < amount) {
+            diff = amount.sub(currentBalance);
+            barn.withdraw(diff);
+        }
+
+        uint256 postBalance = token.balanceOf(address(this));
+
+        require(postBalance >= amount, "BarnPrizePool/insuff-liquidity");
```

Comment:
I think the postBalance should be exactly equal to `amount`

Reply:
Fixed :)